nrf52-hal-common/src/twim.rs

```diff
@@ -230,6 +233,114 @@ impl<T> Twim<T> where T: TwimExt {
+            // waiting for the I2C transaction to end before this stack pointer
+            // becomes invalid, there's nothing wrong here.
+            //
+            // The PTR field is a full 32 bits wide and accepts the full range
+            // of values.
+            unsafe { w.ptr().bits(wr_buffer.as_ptr() as u32) }
+        );
+        self.0.txd.maxcnt.write(|w|
+            // We're giving it the length of the buffer, so no danger of
+            // accessing invalid memory. We have verified that the length of the
+            // buffer fits in an `u8`, so the cast to `u8` is also fine.
+            //
+            // The MAXCNT field is 8 bits wide and accepts the full range of
+            // values.
+            unsafe { w.maxcnt().bits(wr_buffer.len() as _) }
+        );
+
+        // Set up the DMA read
+        self.0.rxd.ptr.write(|w|
+            // We're giving the register a pointer to the stack. Since we're
+            // waiting for the I2C transaction to end before this stack pointer
+            // becomes invalid, there's nothing wrong here.
+            //
+            // The PTR field is a full 32 bits wide and accepts the full range
+            // of values.
+            unsafe { w.ptr().bits(rd_buffer.as_mut_ptr() as u32) }
+        );
+        self.0.rxd.maxcnt.write(|w|
+            // We're giving it the length of the buffer, so no danger of
+            // accessing invalid memory. We have verified that the length of the
+            // buffer fits in an `u8`, so the cast to the type of maxcnt
+            // is also fine.
+            //
+            // Note that that nrf52840 maxcnt is a wider
+            // type than a u8, so we use a `_` cast rather than a `u8` cast.
+            // The MAXCNT field is thus at least 8 bits wide and accepts the
+            // full range of values that fit in a `u8`.
+            unsafe { w.maxcnt().bits(rd_buffer.len() as _) }
+        );
+
+        // Immediately start RX after TX, then stop
+        self.0.shorts.modify(|_r, w|
+            w.lasttx_startrx().enabled()
+             .lastrx_stop().enabled()
+        );
+
+        // Start write operation
+        self.0.tasks_starttx.write(|w|
+            // `1` is a valid value to write to task registers.
+            unsafe { w.bits(1) }
+        );
+
+        // Wait until total operation has ended
+        while self.0.events_stopped.read().bits() == 0 {}
+
+        self.0.events_lasttx.write(|w| w); // reset event
+        self.0.events_lastrx.write(|w| w); // reset event
+        self.0.events_stopped.write(|w| w); // reset event
+        self.0.shorts.write(|w| w);
+
+        let bad_write = self.0.txd.amount.read().bits() != wr_buffer.len() as u32;
+        let bad_read  = self.0.rxd.amount.read().bits() != rd_buffer.len() as u32;
+
+        // Conservative compiler fence to prevent optimizations that do not
+        // take in to account DMA
+        compiler_fence(AcqRel);
```

Comment:
The comment I made in `spim.rs` about a (possibly) missing compiler fence at the method entry applies here too.